src/Microsoft.OData.Core/UriParser/Parsers/SelectExpandOptionParser.cs

```diff
@@ -456,11 +437,188 @@ private List<ExpandTermToken> BuildStarExpandTermToken(PathSegmentToken pathToke
+        private QueryToken ParseInnerFilter()
+        {
+            // advance to the equal sign
+            this.lexer.NextToken();
+            string filterText = this.ReadQueryOption();
+
+            UriQueryExpressionParser filterParser = new UriQueryExpressionParser(this.MaxFilterDepth, enableCaseInsensitiveBuiltinIdentifier);
+            return filterParser.ParseFilter(filterText);
+        }
+
+        /// <summary>
+        /// Parse the orderby option in the select/expand option text.
+        /// </summary>
+        /// <returns>The orderby option for select/expand</returns>
+        private IEnumerable<OrderByToken> ParseInnerOrderBy()
+        {
+            // advance to the equal sign
+            this.lexer.NextToken();
+            string orderByText = this.ReadQueryOption();
+
+            UriQueryExpressionParser orderbyParser = new UriQueryExpressionParser(this.MaxOrderByDepth, enableCaseInsensitiveBuiltinIdentifier);
+            return orderbyParser.ParseOrderBy(orderByText);
+        }
+
+        /// <summary>
+        /// Parse the top option in the select/expand option text.
+        /// </summary>
+        /// <returns>The top option for select/expand</returns>
+        private long? ParseInnerTop()
+        {
+            // advance to the equal sign
+            this.lexer.NextToken();
+            string topText = this.ReadQueryOption();
+
+            // TryParse requires a non-nullable non-negative long.
+            long top;
+            if (!long.TryParse(topText, out top) || top < 0)
+            {
+                throw new ODataException(ODataErrorStrings.UriSelectParser_InvalidTopOption(topText));
+            }
+
+            return top;
+        }
+
+        /// <summary>
+        /// Parse the skip option in the select/expand option text.
+        /// </summary>
+        /// <returns>The skip option for select/expand</returns>
+        private long? ParseInnerSkip()
+        {
+            // advance to the equal sign
+            this.lexer.NextToken();
+            string skipText = this.ReadQueryOption();
+
+            // TryParse requires a non-nullable non-negative long.
+            long skip;
+            if (!long.TryParse(skipText, out skip) || skip < 0)
+            {
+                throw new ODataException(ODataErrorStrings.UriSelectParser_InvalidSkipOption(skipText));
+            }
+
+            return skip;
+        }
+
+        /// <summary>
+        /// Parse the count option in the select/expand option text.
+        /// </summary>
+        /// <returns>The count option for select/expand</returns>
+        private bool? ParseInnerCount()
+        {
+            // advance to the equal sign
+            this.lexer.NextToken();
+            string countText = this.ReadQueryOption();
+            switch (countText)
+            {
+                case ExpressionConstants.KeywordTrue:
+                    return true;
+
+                case ExpressionConstants.KeywordFalse:

```

Comment:
should we support case insensitivity here?

Reply:
Case insensitivity supports for the key word, for the literal of "true" and "false", it's not support case insensitivity. Let me double check it.